src/utils/webrtc/simplewebrtc/peer.js

```diff
@@ -372,22 +400,126 @@ Peer.prototype.end = function() {
 	this.pc.close()
 	this.handleStreamRemoved()
 	this.parent.off('localTrackReplaced', this.handleLocalTrackReplacedBound)
+	this.parent.off('localTrackEnabledChanged', this.handleLocalTrackEnabledChangedBound)
 }
 
 Peer.prototype.handleLocalTrackReplaced = function(newTrack, oldTrack, stream) {
+	this._pendingReplaceTracksQueue.push({ newTrack, oldTrack, stream })
+
+	this._processPendingReplaceTracks()
+}
+
+/**
+ * Process pending replace track actions.
+ *
+ * All the pending replace track actions are executed from the oldest to the
+ * newest, waiting until the previous action was executed before executing the
+ * next one.
+ *
+ * The process may be stopped if the connection is lost, or if a track needs to
+ * be added rather than replaced, which requires a renegotiation. In both cases
+ * the process will start again once the connection is restablished.
+ *
+ * @returns {Promise} a Promise fulfilled when the processing ends; if it was
+ *          completed the resolved value is true, and if it was stopped before
+ *          finishing the resolved value is false.
+ */
+Peer.prototype._processPendingReplaceTracks = function() {
+	if (this._processPendingReplaceTracksPromise) {
+		return this._processPendingReplaceTracksPromise
+	}
+
+	this._processPendingReplaceTracksPromise = this._processPendingReplaceTracksAsync()
+
+	// For compatibility with older browsers "finally" should not be used on
+	// Promises.
+	this._processPendingReplaceTracksPromise.then(() => {
+		this._processPendingReplaceTracksPromise = null
+	}).catch(() => {
+		this._processPendingReplaceTracksPromise = null
+	})
+
+	return this._processPendingReplaceTracksPromise
+}
+
+Peer.prototype._processPendingReplaceTracksAsync = async function() {
+	for (let i = 0; i < this._pendingReplaceTracksQueue.length; i++) {
+		if (this.pc.iceConnectionState === 'new') {
+			// Do not replace the tracks when the connection has not started
+			// yet, as Firefox can get "stuck" and not replace the tracks
+			// even if tried later again once connected.
+			return false
+		}
+
+		const pending = this._pendingReplaceTracksQueue[i]
+
+		try {
+			await this._replaceTrack(pending.newTrack, pending.oldTrack, pending.stream)
+		} catch (exception) {
+			// If the track is added instead of replaced a renegotiation
+			// will be needed, so stop replacing tracks.
+			return false
+		} finally {
+			this._pendingReplaceTracksQueue.splice(i, 1)
```

Comment:
I'm still concerned about a mess here. The call is not protected and it could happen that while we await on the line above, a concurrent call to `handleLocalTrackReplaced` would push a new entry and might mess up by having the index point to something else. Now, maybe this code is already safe still (haven't checked in depth) but might still cause confusion for the reader.

How about we use `this.pendingReplaceTracksQueue.shift()` instead of a loop to remove the items directly as we process them. From what I understand there is no use keeping the array populated in case of error since we abort the whole thing then ? And if we still need it we could re-push or unshift it to the queue then.

Using shift() also means that if concurrently a new item gets added with push(), that item will be added at the end while we still process the items from the beginning.